lua/lspkind/init.lua

```diff
@@ -1,185 +1,309 @@
+  end,
 }
 
--- default true
--- deprecated
-local function opt_with_text(opts)
-  return opts == nil or opts['with_text'] == nil or opts['with_text']
-end
+-- Default options
+H.defaults = {
+  mode = 'symbol_text',
+  symbols = 'default',
+}
 
--- default 'symbol'
-local function opt_mode(opts) 
-    local mode = 'symbol'
-    if opt_with_text(opts) and opts ~= nil and opts['mode'] == nil then
-        mode = 'symbol_text'
-    elseif opts ~= nil and opts['mode'] ~= nil then
-        mode = opts['mode']
-    end
-    return mode
+-- Format symbols
+function H.fmt_kind(list, order, mode, _kind, icon)
+  local kitem = {
+    name = H.get_symbol(list, order, _kind).name,
+    icon = icon,
+  }
+  return H.modes[mode](kitem)
 end
 
--- default 'default'
-local function opt_preset(opts)
-  local preset
-  if opts == nil or opts['preset'] == nil then
-    preset = 'default'
-  else
-    preset = opts['preset']
+-- Define symbols
+function H.set_symbols(ktype, order, symbols, mode)
+  for kind, icon in pairs(symbols) do
+    if ktype[kind] then
+      local idx = ktype[kind]
+      ktype[idx] = H.fmt_kind(ktype, order, mode, kind, icon)
+    end
   end
-  return preset
 end
 
-function lspkind.init(opts)
-  if opts~=nil and opts['with_text'] ~= nil then
-    vim.api.nvim_command("echoerr 'DEPRECATED replaced by mode option.'")
-  end
-  local preset = opt_preset(opts)
-
-  local symbol_map = kind_presets[preset]
-  lspkind.symbol_map = (opts and opts['symbol_map'] and
-                       vim.tbl_extend('force', symbol_map, opts['symbol_map'])) or symbol_map
+-- Setup plugin
+function M.setup(opts)
+  opts = vim.tbl_deep_extend('force', H.defaults, opts or {})
 
-  local symbols = {}
-  local len = kind_len
-  for i = 1, len do
-    local name = kind_order[i]
-    symbols[i] = lspkind.symbolic(name, opts)
+  -- Store symbols, globally accessible
+  if type(opts.symbols) == 'string' then
+    M.symbols = kinds.presets[opts.symbols]
+  else
+    M.symbols = vim.tbl_deep_extend('force', kinds.presets.default, opts.symbols)
   end
 
-  for k,v in pairs(symbols) do
-    require('vim.lsp.protocol').CompletionItemKind[k] = v
-  end
+  H.set_symbols(kinds.list.cmp, kinds.order.cmp, M.symbols, opts.mode)
+  H.set_symbols(kinds.list.sym, kinds.order.sym, M.symbols, opts.mode)
 end
 
-lspkind.presets = kind_presets
-lspkind.symbol_map = kind_presets.default
-
-function lspkind.symbolic(kind, opts)
-  local mode = opt_mode(opts)
-  local formatter = modes[mode]
+-- nvim-cmp
+function M.cmp_format(opts)
+  opts = vim.tbl_deep_extend('force', H.defaults, opts or {})
 
-  -- if someone enters an invalid mode, default to symbol
-  if formatter == nil then
-      formatter = modes['symbol']
-  end
-
-  return formatter(kind)
-end
-
-function lspkind.cmp_format(opts)
-  if opts == nil then
-    opts = {}
-  end
-  if opts.preset or opts.symbol_map then
-    lspkind.init(opts)
+  if opts.symbols then
+    M.setup(opts)
   end
```

Comment:
Although I think it's not your fault here, it already existed, but calling `M.setup` here really sounds weird

Reply:
`lspkind.cmp_format(...)` is a `lspkind.setup(...)` extension

Reply:
No it is not. It creates a template formatting function that can be used in nvim-cmp.

Reply:
I still claim that it is an extension, if you use `lspkind.cmp_format(...)` it is not necessary to define `lspkind.setup(...)`, as it would be duplicating configuration
If `nvim-cmp` is used use `lspkind.cmp_format(...)`, if not, set it with `lspkind.setup(...)`
These functions are not meant to be used at the same time.